src/TLuaInterpreter.cpp

```diff
@@ -842,25 +842,28 @@ int TLuaInterpreter::selectCaptureGroup(lua_State* L)
         lua_pushnumber(L, -1);
         return 1;
     }
-    luaNumOfMatch--; //we want capture groups to start with 1 instead of 0
-    if (luaNumOfMatch < static_cast<int>(host.getLuaInterpreter()->mCaptureGroupList.size())) {
+    // We want capture groups to start with 1 instead of 0 so predecrement
+    // luaNumOfMatch :
+    if (--luaNumOfMatch < static_cast<int>(host.getLuaInterpreter()->mCaptureGroupList.size())) {
         TLuaInterpreter* pL = host.getLuaInterpreter();
         auto iti = pL->mCaptureGroupPosList.begin();
         auto its = pL->mCaptureGroupList.begin();
+        int begin = *iti;
+        std::string& s = *its;
 
         for (int i = 0; iti != pL->mCaptureGroupPosList.end(); ++iti, ++i) {
+            begin = *iti;
             if (i >= luaNumOfMatch) {
                 break;
             }
         }
         for (int i = 0; its != pL->mCaptureGroupList.end(); ++its, ++i) {
+            s = *its;
             if (i >= luaNumOfMatch) {
                 break;
             }
         }
 
-        int begin = *iti;
-        std::string& s = *its;
```

Comment:
Almost everything before this could be replaced by:

```
int begin = std::advance(pL->mCaptureGroupPosList.begin(), luaNumOfMatch);
std::string& s = std::advance(pL->mCaptureGroupList.begin(), luaNumOfMatch);
```

Though, the if statement should check that both lists are long enough.

Reply:
That would make it a lot more comprehensible! Thanks for the tip

Reply:
Not sure I can get my head around that - I was just trying to fix the iterator error but if you have a better form please push a PR out to this on my repo...